### ISSUE TITLE
Varied results of `g.apply_edges(fn.u_mul_v('h', 'h', 'w_new'))`.

### ISSUE THREAD

**Issue by jzhang-github** [dmlc/dgl#6378]
## 🐛 Bug

Varied results of `g.apply_edges(fn.u_mul_v('h', 'h', 'w_new'))`.

In most cases, the result is (dependent on the random variables):
```python
tf.Tensor(
[[4.2640194e-01 1.6625892e-01 6.2569749e-01 5.5540627e-01 1.0956249e-01
  4.1180290e-03 1.6604912e-01 5.1894283e-01 1.9427264e-02 7.9354626e-01]
 [5.3002572e-01 7.1355954e-02 3.2891789e-01 6.8925512e-01 3.7086263e-02
  1.1190422e-02 1.3854278e-03 2.8031564e-01 3.5215333e-02 2.3537005e-01]
 [4.8363465e-01 2.0609375e-04 2.4031331e-01 3.8903135e-01 4.5351457e-02
  5.4170787e-01 3.4699936e-03 5.0742984e-01 2.7762973e-01 3.3470146e-02]], shape=(3, 10), dtype=float32)
```

If I run `g_test.apply_edges(fn.u_mul_v('h', 'h', 'w_new'))` multiple times without refreshing random variables, I sometime have the following outputs:
```python
tf.Tensor(
[[0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]], shape=(3, 10), dtype=float32)
```

## To Reproduce

1. Prepare the following code in a console. I use the `Spyder`.
   ```python
   import dgl
   import dgl.function as fn
   import tensorflow as tf

   src_ids = tf.constant([2, 3, 4])
   dst_ids = tf.constant([1, 2, 3])
   g = dgl.graph((src_ids, dst_ids))
   g.ndata['h'] = tf.random.uniform((g.num_nodes(), 10)) # each node has feature size 10
   g.edata['w'] = tf.random.uniform((g.num_edges(), 1))  # each edge has feature size 1
   g.update_all(fn.copy_u('h', 'm'), fn.sum('m', 'h_sum'))
   g.update_all(fn.u_mul_e('h', 'w', 'm'), fn.max('m', 'h_max'))

   # The following lines are key to reproduce the error.
   g.apply_edges(fn.u_mul_v('h', 'h', 'w_new'))
   print(g.edata['w_new'])
   ```

2. Run the code for the first time, and the random variables will be initialized.
3. Run the last two lines without refreshing the code randomness, you may have different outputs. I have a tf.tensor with all zeros.

## Environment

 - DGL Version: 1.1.1+cu118
 - Backend Library & Version: Tensorflow 2.13.0
 - OS (e.g., Linux): Linux
 - How you installed DGL: maybe conda `conda install -c dglteam/label/cu118 dgl`
 - Build command you used (if compiling from source): 
 - Python version: 3.10
 - CUDA/cuDNN version (if applicable): CUDA Version: 12.2
 - GPU models and configuration: NVIDIA RTX A6000
 - Any other relevant information: not applicable

**Comment by frozenbugs** [comment]
Hi @jzhang-github , I was not able to reproduce the issue, can you double check?
https://colab.research.google.com/drive/13HhUKqZvdUwOwqjQ0dX_x5RddDWK-PGV#scrollTo=HszKugZRKoUI

**Comment by jzhang-github** [comment]
> Hi @jzhang-github , I was not able to reproduce the issue, can you double check? https://colab.research.google.com/drive/13HhUKqZvdUwOwqjQ0dX_x5RddDWK-PGV#scrollTo=HszKugZRKoUI

Hi @frozenbugs , thank you for your reply. I included `cuda-11.8` libraries into `PATH` and `LD_LIBRARY_PATH` , then the strange disappeared. I guess this issue may be caused by my pure `cuda-12.2 `.

**Comment by jzhang-github** [comment]
I recently did more tests on this issue.

The issue may be caused by running parallel processes on the same card. If you use the GPU memory close to 100% , this issue persists.

I really hope the developers could indicate the source of this issue: wrongdoing of user, cuda, tensorflow memory growth, or dgl?

### To reproduce
The code:
```python
import dgl
import dgl.function as fn
import tensorflow as tf

# Source nodes for edges (2, 1), (3, 2), (4, 3)
src_ids = tf.constant([2, 3, 4])
# Destination nodes for edges (2, 1), (3, 2), (4, 3)
dst_ids = tf.constant([1, 2, 3])
g = dgl.graph((src_ids, dst_ids))
# g_test = ... # create a DGLGraph
g.ndata['h'] = tf.random.uniform((g.num_nodes(), 10)) # each node has feature size 10
g.edata['w'] = tf.random.uniform((g.num_edges(), 1))  # each edge has feature size 1
# collect features from source nodes and aggregate them in destination nodes
g.update_all(fn.copy_u('h', 'm'), fn.sum('m', 'h_sum'))
# multiply source node features with edge weights and aggregate them in destination nodes
g.update_all(fn.u_mul_e('h', 'w', 'm'), fn.max('m', 'h_max'))
# compute edge embedding by multiplying source and destination node embeddings


for i in range(20):
    g.apply_edges(fn.u_mul_v('h', 'h', 'w_new'))
    print('========================')
    print(g.edata['w_new'])
```

My output:
```python
2023-09-27 22:00:14.250049: I tensorflow/core/platform/cpu_feature_guard.cc:182] This TensorFlow binary is optimized to use available CPU instructions in performance-critical operations.
To enable the following instructions: AVX2 FMA, in other operations, rebuild TensorFlow with the appropriate compiler flags.
2023-09-27 22:00:14.897858: W tensorflow/compiler/tf2tensorrt/utils/py_utils.cc:38] TF-TRT Warning: Could not find TensorRT
2023-09-27 22:00:16.726146: W tensorflow/core/common_runtime/gpu/gpu_bfc_allocator.cc:47] Overriding orig_value setting because the TF_FORCE_GPU_ALLOW_GROWTH environment variable is set. Original config value was 0.
2023-09-27 22:00:16.726217: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1639] Created device /job:localhost/replica:0/task:0/device:GPU:0 with 6146 MB memory:  -> device: 0, name: NVIDIA RTX A6000, pci bus id: 0000:41:00.0, compute capability: 8.6
========================
tf.Tensor(
[[0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]], shape=(3, 10), dtype=float32)
========================
tf.Tensor(
[[0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]], shape=(3, 10), dtype=float32)
========================
tf.Tensor(
[[0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]], shape=(3, 10), dtype=float32)
========================
tf.Tensor(
[[0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]], shape=(3, 10), dtype=float32)
========================
tf.Tensor(
[[0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]], shape=(3, 10), dtype=float32)
========================
tf.Tensor(
[[0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]], shape=(3, 10), dtype=float32)
========================
tf.Tensor(
[[1.0253576e-02 1.3075180e-01 5.7482266e-01 6.7359045e-02 1.5899391e-01
  2.2390601e-01 9.7890697e-02 6.2681255e-03 1.2859498e-01 3.4314601e-04]
 [2.8556930e-03 6.2411278e-01 6.4070112e-01 3.4034211e-02 2.6684648e-01
  4.8859417e-02 2.2112578e-03 2.0315725e-01 1.0968258e-01 2.8247829e-04]
 [4.0815558e-02 7.5615031e-01 3.8724628e-01 6.1505869e-02 6.2921041e-01
  1.5420045e-01 2.1904839e-04 2.9457828e-02 4.9425358e-01 1.8605481e-01]], shape=(3, 10), dtype=float32)
========================
tf.Tensor(
[[1.0253576e-02 1.3075180e-01 5.7482266e-01 6.7359045e-02 1.5899391e-01
  2.2390601e-01 9.7890697e-02 6.2681255e-03 1.2859498e-01 3.4314601e-04]
 [2.8556930e-03 6.2411278e-01 6.4070112e-01 3.4034211e-02 2.6684648e-01
  4.8859417e-02 2.2112578e-03 2.0315725e-01 1.0968258e-01 2.8247829e-04]
 [4.0815558e-02 7.5615031e-01 3.8724628e-01 6.1505869e-02 6.2921041e-01
  1.5420045e-01 2.1904839e-04 2.9457828e-02 4.9425358e-01 1.8605481e-01]], shape=(3, 10), dtype=float32)
========================
tf.Tensor(
[[1.0253576e-02 1.3075180e-01 5.7482266e-01 6.7359045e-02 1.5899391e-01
  2.2390601e-01 9.7890697e-02 6.2681255e-03 1.2859498e-01 3.4314601e-04]
 [2.8556930e-03 6.2411278e-01 6.4070112e-01 3.4034211e-02 2.6684648e-01
  4.8859417e-02 2.2112578e-03 2.0315725e-01 1.0968258e-01 2.8247829e-04]
 [4.0815558e-02 7.5615031e-01 3.8724628e-01 6.1505869e-02 6.2921041e-01
  1.5420045e-01 2.1904839e-04 2.9457828e-02 4.9425358e-01 1.8605481e-01]], shape=(3, 10), dtype=float32)
========================
tf.Tensor(
[[0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]], shape=(3, 10), dtype=float32)
========================
tf.Tensor(
[[0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]], shape=(3, 10), dtype=float32)
========================
tf.Tensor(
[[0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]], shape=(3, 10), dtype=float32)
========================
tf.Tensor(
[[0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]], shape=(3, 10), dtype=float32)
========================
tf.Tensor(
[[0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]], shape=(3, 10), dtype=float32)
========================
tf.Tensor(
[[0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]], shape=(3, 10), dtype=float32)
========================
tf.Tensor(
[[0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]], shape=(3, 10), dtype=float32)
========================
tf.Tensor(
[[0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]], shape=(3, 10), dtype=float32)
========================
tf.Tensor(
[[0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]], shape=(3, 10), dtype=float32)
========================
tf.Tensor(
[[0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]], shape=(3, 10), dtype=float32)
========================
tf.Tensor(
[[1.0253576e-02 1.3075180e-01 5.7482266e-01 6.7359045e-02 1.5899391e-01
  2.2390601e-01 9.7890697e-02 6.2681255e-03 1.2859498e-01 3.4314601e-04]
 [2.8556930e-03 6.2411278e-01 6.4070112e-01 3.4034211e-02 2.6684648e-01
  4.8859417e-02 2.2112578e-03 2.0315725e-01 1.0968258e-01 2.8247829e-04]
 [4.0815558e-02 7.5615031e-01 3.8724628e-01 6.1505869e-02 6.2921041e-01
  1.5420045e-01 2.1904839e-04 2.9457828e-02 4.9425358e-01 1.8605481e-01]], shape=(3, 10), dtype=float32)
```

**Comment by frozenbugs** [comment]
Hi @jzhang-github , thanks for digging into the details. Is this TF only issue or also reproducible on Pytorch.

As we announced in our slack channel: After DGL 2.0 (ETA: 2023Q4), we'll be prioritizing PyTorch backend due to resource constraints, saying goodbye to MXNet and TensorFlow support.

**Comment by jzhang-github** [comment]
> Hi @jzhang-github , thanks for digging into the details. Is this TF only issue or also reproducible on Pytorch.
> 
> As we announced in our slack channel: After DGL 2.0 (ETA: 2023Q4), we'll be prioritizing PyTorch backend due to resource constraints, saying goodbye to MXNet and TensorFlow support.

Hi, this issue is not detected on the PyTorch backend.

**Comment by frozenbugs** [comment]
Thanks for the clarification, I will close this issue for now.
If you're interested in investigating this further and fixing it, feel free to reopen the issue and modify the related DGL-TF code. We're happy to help you get your code merged.
